extensions/identity-hub-client/src/main/java/org/eclipse/dataspaceconnector/identityhub/client/IdentityHubClientImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ *  Copyright (c) 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.identityhub.client;
+
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.ws.rs.core.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.eclipse.dataspaceconnector.identityhub.models.Descriptor;
+import org.eclipse.dataspaceconnector.identityhub.models.MessageRequestObject;
+import org.eclipse.dataspaceconnector.identityhub.models.RequestObject;
+import org.eclipse.dataspaceconnector.identityhub.models.ResponseObject;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.eclipse.dataspaceconnector.identityhub.models.WebNodeInterfaceMethod.COLLECTIONS_QUERY;
+import static org.eclipse.dataspaceconnector.identityhub.models.WebNodeInterfaceMethod.COLLECTIONS_WRITE;
+
+public class IdentityHubClientImpl implements IdentityHubClient {
+
+    private final OkHttpClient httpClient;
+    private final ObjectMapper objectMapper;
+
+    public IdentityHubClientImpl(OkHttpClient httpClient, ObjectMapper objectMapper) {
+        this.httpClient = httpClient;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public Collection<VerifiableCredential> getVerifiableCredentials(String hubBaseUrl) {
+        ResponseObject responseObject;
+        try {
+            Response response = httpClient.newCall(
+                        new Request.Builder()
+                                .url(hubBaseUrl)
+                                .post(buildRequestBody(COLLECTIONS_QUERY.getName()))
+                                .build())
+                .execute();
+
+            responseObject = objectMapper.readValue(response.body().byteStream(), ResponseObject.class);
+
+        } catch (IOException e) {
+            throw new EdcException(e);
+        }
+
+        Collection<VerifiableCredential> entries = responseObject.getReplies().stream()
+                .findFirst()
+                .orElseThrow(() -> new EdcException("Invalid response"))
+                .getEntries().stream()
+                    .map(e -> objectMapper.convertValue(e, VerifiableCredential.class)).collect(Collectors.toList());
```

Comment:
if here we are assuming that the hub only contains `VerifiableCredential` objects, shouldn't we check in `CollectionsWriteProcessor` that incoming data actually deserialises to that class before we persist it?

Reply:
Thanks for the recommendation.
Now if the object is not a VerifiableCredential it is not saved, and it will returned a MALFORMED_MESSAGE message status.